prometheus_keys.lua

```diff
@@ -0,0 +1,118 @@
+-- Storage to keep track of used keys. Allows to atomically create, delete
+-- and list keys. The keys are synchronized between nginx workers
+-- using ngx.shared.dict. The whole purpose of this module is to avoid
+-- using ngx.shared.dict:get_keys (see https://github.com/openresty/lua-nginx-module#ngxshareddictget_keys),
+-- which blocks all workers and therefore it shouldn't be used with large
+-- amounts of keys.
+
+local KeyIndex = {}
+KeyIndex.__index = KeyIndex
+
+function KeyIndex.new(shared_dict, prefix)
+  local self = setmetatable({}, KeyIndex)
+  self.dict = shared_dict
+  self.key_prefix = prefix .. "key_"
+  self.delete_count = prefix .. "delete_count"
+  self.key_count = prefix .. "key_count"
+  self.last = 0
+  self.deleted = 0
+  self.keys = {}
+  self.index = {}
+  return self
+end
+
+-- Loads new keys that might have been added by other workers since last sync.
+function KeyIndex:sync()
+  local delete_count = self.dict:get(self.delete_count) or 0
+  local N = self.dict:get(self.key_count) or 0
+  if self.deleted ~= delete_count then
+    -- Some other worker deleted something, lets do a full sync.
+    self:sync_range(0, N)
+    self.deleted = delete_count
+  elseif N ~= self.last then
+    -- Sync only new keys, if there are any.
+    self:sync_range(self.last, N)
+  end
+  return N
```

Comment:
Just noticed that `self.last` does not get updated here, which means every sync is a full sync.

Reply:
Fixed in 84dd82ec4f0542a5245a1fac4fbb410a5c1bcbba